docs: improve formatted output for oras discover (#1625)
oras-project/oras: docs/proposals/formatted-output.md

    └── sha256:50fd0dc107d84b5e7b402688000a7ed3aaf8a2692d5cb74da5277fa3c4cecf15
```

> [!NOTE]

> > The `--format` flag replaces the existing `--output` flag. The `--output` has been marked as "deprecated" in ORAS v1.2.0 and will be removed in future releases.

Reviewer: Currently, the `--output` flag is still visible in the help doc. I think we can just hide it. Should we mention this in the spec?

Author: Or we could consider moving this discussion to an issue and create a doc to elaborate on the feature lifecycle management.

Author: @wangxiaoxuan273 @Wwwsylvia @oras-project/oras-cli I created an issue to track this discussion https://github.com/oras-project/oras-www/issues/460. We can discuss it in the issue and align on the deprecation policy, then make the changes to corresponding deprecated flags. I will mark this comment as resolved.

> > The `--format` flag will replace the existing `--output` flag. The `--output` SHOULD be marked as "deprecated" in ORAS v1.2.0 and will be removed in future releases.

Reviewer 2: ```suggestion
> The `--format` flag replaces the existing `--output` flag. The `--output` has been marked as "deprecated" in ORAS v1.2.0 and will be removed in future releases.
```

> If the referrers have child referrers, ORAS SHOULD show the manifest content of the subject image and all referrers recursively in all formatted outputs (`tree`, `json`). It ensures data consistency of different data formats in the output.

Reviewer: > all referrers recursively in all formatted outputs (`tree`, `json`)

Should we also mention the `go-template` format?

> If the referrers have child referrers, ORAS SHOULD show the manifest content of the subject image and all referrers recursively in all formatted outputs (tree, JSON). It ensures data consistency of different data formats in the output.

Reviewer 2: ```suggestion
If the referrers have child referrers, ORAS SHOULD show the manifest content of the subject image and all referrers recursively in all formatted outputs (`tree`, `json`). It ensures data consistency of different data formats in the output.
```

#### JSON output of `oras discover`

> When showing the subject image and all referrers' manifests recursively in formatted JSON output, the following fields should be returned:

Reviewer: nit: how about:

> When showing the subject image and all referrers' manifests recursively in a formatted JSON output, include the following fields:

      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:a811606b09341bab4bbc0a4deb2c0cb709ec9702635cbe2d36b77d58359ec046",
      "size": 747,
      "annotations": {
        "org.opencontainers.image.created": "2024-01-18T18:12:41Z"
      },
      "artifactType": "application/vnd.in-toto+json",
      "referrers": [
        {
          "reference": "localhost:5000/kubernetes/kubectl@sha256:04723fd7d00df77c6f226b907667396554bf9418dc48a7a04feb5ff24aa0b9ec",
          "mediaType": "application/vnd.oci.image.manifest.v1+json",
          "digest": "sha256:04723fd7d00df77c6f226b907667396554bf9418dc48a7a04feb5ff24aa0b9ec",
          "size": 1080,
          "annotations": {
            "io.cncf.notary.x509chain.thumbprint#S256": "[430ecf0685f8018443f8418f5d7134b146f28862116114925713635d5703fb69,9b1894f223d934cbd6575af3c6e1f6096b9221a7da132185f5a5cdc92235b5dc,23ffe2b8bdb9a1711515d4cffda04bc7f793d513c76c243f1020507d8669b7db]",
            "org.opencontainers.image.created": "2024-01-18T18:20:09Z"
          },
          "artifactType": "application/vnd.cncf.notary.signature"
        }
      ]
    }
  ]
```

#### Table format of `oras discover` output

> The `oras discover --format table` command currently lists only direct referrers in a table format. However, the table view has limitations—it does not effectively represent the hierarchical artifact reference relationships of all referrers. Users can't differentiate the reference relationship between each referrer in a table format output. See current sample output as follows:

Reviewer: nit: "however" might be confusing here, as "The `oras discover --format table` command currently lists only direct referrers in a table format" implies that it cannot represent the hierarchical information.
How about:

```suggestion
The `oras discover --format table` command currently lists only direct referrers in a table format.  This table view has limitations: it does not effectively represent the hierarchical artifact reference relationships of all referrers. Users can't differentiate the reference relationship between each referrer in a table format output. See current sample output as follows:
```

```console
$ oras discover localhost:5000/kubectl:v1.29.1 --format table
Discovered 4 artifacts referencing localhost:5000/kubectl:v1.29.1
Digest: sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01

Artifact Type                                  Digest
application/vnd.microsoft.artifact.lifecycle   sha256:325129be79f416fe11a9ec44233cfa57f5d89434e6d37170f97e48f7904983e3
application/vnd.cncf.notary.signature          sha256:f520330e9f43c05859c532e67a25c9c765b144782ae7b872656192c27fd4e2dd
application/vnd.in-toto+json                   sha256:a811606b09341bab4bbc0a4deb2c0cb709ec9702635cbe2d36b77d58359ec046
application/vnd.cncf.notary.signature          sha256:f2098a230b6311edeb44ab2d6e5789372300d9b98be34c4d9477d3b9638a3bb1
```

> In contrast, the tree view offers a more structured and human-readable output by visually displaying relationships of each referrer. Given this advantage, the table view no longer provides significant value to users. This document proposes deprecating the `--format table` option. When user specifies `--format table` flag with `oras discover`, the warning message will be printed out as follows:

Reviewer: nit:
```suggestion
In contrast, the tree view offers a more structured and human-readable output by visually displaying relationships of each referrer. Given this advantage, the table view no longer provides significant value to users. This document proposes deprecating the `--format table` option. When user specifies `--format table` flag with `oras discover`, the warning message should be printed out as follows:
```

```bash
oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01 --format tree --depth 3
```

```console
localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01
├── application/vnd.oci.artifact.lifecycle
│   └── sha256:325129be79f416fe11a9ec44233cfa57f5d89434e6d37170f97e48f7904983e3
│       └── application/vnd.cncf.notary.signature
│           └── sha256:f520330e9f43c05859c532e67a25c9c765b144782ae7b872656192c27fd4e2dd
│               └── vnd/test-annotations
│                   └── sha256:d2cb66a53e4d77488df1f15701554ebb11ffa1cf6eb90f79afa33d3b172f11d2
└── application/vnd.in-toto+json
    └── sha256:a811606b09341bab4bbc0a4deb2c0cb709ec9702635cbe2d36b77d58359ec046
        └── application/vnd.cncf.notary.signature
            └── sha256:04723fd7d00df77c6f226b907667396554bf9418dc48a7a04feb5ff24aa0b9ec
                └── vnd/test-annotations
                    └── sha256:d2cb66a53e4d77488df1f15701554ebb11ffa1cf6eb90f79afa33d3b172f11d2
```

## FAQ

**Q:** Why choose to use `--format` flag to enable JSON formatted output instead of extending the existing `--output` flag?

> **A:** ORAS follows [GNU](https://www.gnu.org/prep/standards/html_node/Option-Table.html#Option-Table) design principles. ORAS uses `--output` to specify a file or directory content should be created within and `--format` to format the output into JSON or using the given Go template. Popular tools, like Docker, Podman, and Skopeo also follow this design principle within their formatted output feature.

Reviewer 3: nit: `Q` and `A` should be in different lines. Currently in preview they are in the same line. Consider adding an extra blank line before `A`.

Reviewer: +1

    └── sha256:50fd0dc107d84b5e7b402688000a7ed3aaf8a2692d5cb74da5277fa3c4cecf15
```

> [!NOTE]

> > The `--output` flag will replaced by `--format` flag. The `--output` SHOULD be marked as "deprecated" in ORAS with warning message and will be removed in future releases. 

Reviewer: nit: 
```suggestion
> The `--output` flag will be replaced by the `--format` flag. The `--output` flag SHOULD be marked as "deprecated" in ORAS with a warning message and will be removed in future releases. 
```

> View an artifact's referrers manifest in pretty JSON output. The following fields should be outputted:

Reviewer 3: After #1403 is completed, the subject manifest will also be in the output. This should be reflected in this spec. I have named the struct field temporarily as `subject`, we can discuss about it.

Author: We will keep what this spec proposed per our discussion and won't introduce an additional `subject` field since it might confuse users.

View an artifact's referrers manifest in pretty JSON output. The following fields should be outputted:

> 

Reviewer 3: Shouldn't you still include the info for the subject here? The `reference`, `mediaType`, `size`, `digest` fields of the subject manifest before the `referrers` field?

Author: Updated

├── application/vnd.oci.artifact.lifecycle
│   └── sha256:325129be79f416fe11a9ec44233cfa57f5d89434e6d37170f97e48f7904983e3
│       └── application/vnd.cncf.notary.signature
│           └── sha256:f520330e9f43c05859c532e67a25c9c765b144782ae7b872656192c27fd4e2dd
└── application/vnd.in-toto+json
    └── sha256:a811606b09341bab4bbc0a4deb2c0cb709ec9702635cbe2d36b77d58359ec046
        └── application/vnd.cncf.notary.signature
            └── sha256:04723fd7d00df77c6f226b907667396554bf9418dc48a7a04feb5ff24aa0b9ec
```

When showing the subject image and all referrers' manifests recursively in a pretty JSON output, the following JSON output should be returned:

```bash
oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01 --format json
```

```json
  "reference": "localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01",
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "digest": "sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01",
  "size": 1788,

>   "referrers": [

Reviewer 3: This UX is different from my current implementation in #1649 and I think it's a good design! We may need to discuss on the choice. This part still needs to be updated if we choose this UX.

For example, assume there is a sample image with four-level referrers, show referrers within the thrid level only in a tree view:

```bash
oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01 --format tree --depth 3
```

```console
localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01
├── application/vnd.oci.artifact.lifecycle
│   └── sha256:325129be79f416fe11a9ec44233cfa57f5d89434e6d37170f97e48f7904983e3
│       └── application/vnd.cncf.notary.signature
│           └── sha256:f520330e9f43c05859c532e67a25c9c765b144782ae7b872656192c27fd4e2dd
│               └── vnd/test-annotations
│                   └── sha256:d2cb66a53e4d77488df1f15701554ebb11ffa1cf6eb90f79afa33d3b172f11d2
└── application/vnd.in-toto+json
    └── sha256:a811606b09341bab4bbc0a4deb2c0cb709ec9702635cbe2d36b77d58359ec046
        └── application/vnd.cncf.notary.signature
            └── sha256:04723fd7d00df77c6f226b907667396554bf9418dc48a7a04feb5ff24aa0b9ec
                └── vnd/test-annotations
                    └── sha256:d2cb66a53e4d77488df1f15701554ebb11ffa1cf6eb90f79afa33d3b172f11d2
```

> 

Reviewer 3: We are going to deprecate the table view. This should be included in the spec.

Author: Updated @wangxiaoxuan273

> [!NOTE]
> The `--format` flag will replace the existing `--output` flag. The `--output` will be marked as "deprecated" in ORAS v1.2.0 and will be removed in the future releases.

If the referrers have associated referrers, ORAS should be able to show the manifest content of the subject image and all referrers. 

When showing the subject image and all referrers' manifests recursively in a tree view output, the following fields should be returned:

- `reference`: full reference by digest of the subject image
- `mediaType`: media type of the subject image
- `digest`: digest of the subject image
- `size`: subject image size in bytes

> - `manifests`: the list of referrers' manifest

Reviewer 2: What if a manifest is not JSON in the future?

  - `reference`: full reference by digest of the referrer
  - `mediaType`: media type of the referrer
  - `digest`: digest of the referrer
  - `size`: referrer file size in bytes
  - `artifactType`: artifact type of a referrer
  - `annotations`: contains arbitrary metadata in a referrer

>   - `referrerManifests`: the list of referrers' manifest

Reviewer 4: What about `manifests` or `referrers`

Reviewer: In terms of consistency, `manifests` actually makes more sense as the schema is the same, just that users may not understand what does "manifests" mean. Maybe we should have called the root manifests `referrers` in the first place😂...

Author: We will use `referrers` as the field name in those two places based on the community discussion.

>   - `referrerManifests`: the list of referrers' manifest

Reviewer 2: What if this field becomes a field of the manifest?

Reviewer 4: Yes the ORAS manifest output schema can have an optional referrers which is a self referencing type.

Author: Updated

> For example, when there are two refferers lifecycle metadata and in-toto attestation associated with a sample image, the signatures are associated with these two files respectively. The output in a tree view will be:

Reviewer 2: There are typos: `refferers` --> `referrers`.

Author: Fixed typo

```bash
oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01
```

```console
localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01
├── application/vnd.oci.artifact.lifecycle
│   └── sha256:325129be79f416fe11a9ec44233cfa57f5d89434e6d37170f97e48f7904983e3
│       └── application/vnd.cncf.notary.signature
│           └── sha256:f520330e9f43c05859c532e67a25c9c765b144782ae7b872656192c27fd4e2dd
└── application/vnd.in-toto+json
    └── sha256:a811606b09341bab4bbc0a4deb2c0cb709ec9702635cbe2d36b77d58359ec046
        └── application/vnd.cncf.notary.signature
            └── sha256:04723fd7d00df77c6f226b907667396554bf9418dc48a7a04feb5ff24aa0b9ec
```

When showing the subject image and all referrers' manifests recursively in a pretty JSON output, the following JSON output should be returned:

```bash

> oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01 --format json --recursive

Reviewer 4: +1 to @Wwwsylvia's comment - `recursive` is default in tree and not in -format json sounds ood. 
flag defaults changing depending on output formats seems really odd.

Author: Changed to show all referrers recursively by default and removed the flag based on our discussion.

> oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01 --format json --recursive

Reviewer 2: As @Wwwsylvia pointed out, the `--recursive` flag is newly introduced and may be inconsistent for other formats.

Reviewer 2: Instead, we can introduce a `--depth` flag.

Author: Added a new section to introduce a new `--depth` flag

> oras discover localhost:5000/kubernetes/kubectl@sha256:bece4f4746a39cb39e38451c70fa5a1e5ea4fa20d4cca40136b51d9557918b01 --format json --recursive

Reviewer 5: Do we want to set a limit on the recursive depth?

```

> ```

Reviewer 2: nit: this code block is not a valid json

Author: Added JSON code block
